### PR TITLE
Aimed Neurotoxin Spit

### PR DESCRIPTION
**Pull request by GunHog**
- Neurotoxin spit is now an aimed projectile.
- Research analyzer action button updated. (To avoid conflicts with my own PRs)
- Framework for adding more aimed things later

**This PR on behalf of @WJohn! His sprites as well!**

Spit now works by a toggle. Turn it on, and your left-clicks will fire spit. Turn it off to return to normal interactions. If this system is favored, it will provide a framework for moving over other things, such as the Wizard's Fireball spell. (Bad aim with Fireball is one of the largest causes of self-inflicted wizard deaths!)

:cl:
Name: Gun Hog
rcsadd: Alien (xeno) neurotoxic spit is now a click-to-aim projectile.
/:cl:

This may not be the optimal method of implementation. Experienced coder feedback requested.

Corresponding fix PR here: https://github.com/tgstation/tgstation/pull/18134

screenshot 2016-06-01 13 42 40

Sorry for not having a proper .gif.

### CODE REVIEW

File: code/modules/mob/living/carbon/carbon_defines.dm [hunk 1]
```diff
@@ -32,3 +32,5 @@
 	var/tinttotal = 0	// Total level of visualy impairing items
 
 	var/list/bodyparts = list() //Gets filled up in the constructor (New() proc in human.dm and monkey.dm)
+
+	var/obj/effect/proc_holder/ranged_ability //Any ranged ability the mob has, as a click override
```
Comment (coiax) [inline code/modules/mob/living/carbon/carbon_defines.dm]: So only carbons get this? Where are my drone fireballs? :( :( :(

Comment (GunHog) [inline code/modules/mob/living/carbon/carbon_defines.dm]: ...Why do you have to do this to me? I will _try_ to move this up to /living if it does not cause too much trouble.

**Comment by GunHog** [comment]
I moved it up to /living. I will need to do more extensive testing, but for now, it seems okay.

**Comment by PKPenguin321** [comment]
Xeno buffs get out

**Comment by WJohn** [comment]
:+1: 
Thank you so much! This'll be a lot of fun to play with.

File: code/modules/mob/living/living_defines.dm [hunk 2]
```diff
@@ -68,3 +68,5 @@
 
 	var/mob/living/mental_dominator //The person controlling the mind of this person, if applicable
 	var/mob/living/mind_control_holder/mind_control_holder //If the mob is being mind controlled, where their old mind is stored (check clock_mobs.dm)
+
+	var/obj/effect/proc_holder/ranged_ability //Any ranged ability the mob has, as a click override
```
Comment (RemieRichards) [inline code/modules/mob/living/living_defines.dm]: why not just use the click override system?

File: code/_onclick/other_mobs.dm [hunk 3]
```diff
@@ -41,6 +41,14 @@
 /mob/living/carbon/RestrainedClickOn(atom/A)
 	return 0
 
+//Ranged override abilities for carbons
+/mob/living/ClickOn(atom/A, params)
+	if(ranged_ability && !stat)
+		ranged_ability.use_ability(A, src, params) //Spit at something, use a fireball, etc.
```
Comment (RemieRichards) [inline code/_onclick/other_mobs.dm]: click intercept system takes these exact same arguments, but as ```src, params, A```, use that system.

Comment (GunHog) [inline code/_onclick/other_mobs.dm]: I do not know how...and is that not client based?

Comment (RemieRichards) [inline code/_onclick/other_mobs.dm]: You just define ```/THING/proc/InterceptClickOn(clicker, params, clicked_thing)```
It also is client based, but use your imagination, add a var to mobs that allow them to pass these intercepts to clients during ```Login()```

**Comment by GunHog** [comment]
I have swapped over to the click intercept system based on clients as ordered, but I am going to be honest and say that I have near zero confidence in the integrity of this code. I do not know how to test it properly (beyond making sure it works and clears the intercepts on logout/ghosting), what problems this may cause with whatever used it before this, and I am not sure if it is safe to port the other abilities (such as Wizard fireball) to this system either.

**Comment by coiax** [comment]
@GunHog I'm literally being serious when I say it's not like Fireball could get any MORE dangerous to the user at present.

**Comment by GunHog** [comment]
The code, I mean. The system I am using is new to me, so I do not know if it is safe, codewise, to port anything else to it until alien spit is fully tested and confirmed bug free. I am not referring to balance or gameplay implications in this case. If I am allowed to port Fireball to this new system, then it will be in a separate PR where drama and balance debate can be contained.

File: code/modules/mob/living/living_defines.dm (hunk 2 above)
Comment (phil235) [inline code/modules/mob/living/living_defines.dm]: nnnnghhh adding a var to all living mob just for a tiny feature that only xeno use....nnghhghg

Comment (GunHog) [inline code/modules/mob/living/living_defines.dm]: Humans can use it, and I plan to port other things, such as the Fireball spell to it, which any /living mob can use. It is built for the future!

**Comment by optimumtact** [comment]
you should arrange a test merge

**Comment by GunHog** [comment]
>you should arrange a test merge

I would prefer to have @RemieRichards give the okay, but a test merge to confirm that it does not ruin players could work too I suppose

**Comment by Ultimate-Chimera** [comment]
Oh boy, a test merge? I wanna test!

**Comment by GunHog** [comment]
Cooldown added at request of WJ.

**Comment by GunHog** [comment]
Conflicts resolved, it is ready again.

**Comment by optimumtact** [comment]
This is a great change @GunHog, thanks :+1:

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 4]
```diff
@@ -190,26 +190,55 @@ Doesn't work on other aliens/AI.*/
...
-	A.yo = U.y - T.y
-	A.xo = U.x - T.x
+	A.preparePixelProjectile(target, get_turf(target), user, params)
 	A.fire()
 	user.newtonian_move(get_dir(U, T))
+	user.adjustPlasma(-p_cost)
 
 	return 1
 
+/obj/effect/proc_holder/alien/on_lose(mob/living/carbon/user)
```
Comment (ChangelingRain) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: shouldn't this be `/obj/effect/proc_holder/alienneurotoxin`?

Comment (GunHog) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: Good catch, standby

**Pull request merged**